### PR TITLE
Use .to to convert new tensors in new_tensor

### LINKED ISSUE #14732: torch.tensor throws a Runtime exception when default tensor type is CUDA
## 🐛 Bug

Ref. to https://github.com/uber/pyro/pull/1620. This seems to be a recent issue in the pytorch nightly release where converting a numpy float64 array to a torch.float32 tensor throws an error if the default tensor type is CUDA. This seemed to work fine earlier. I am not sure if this is a regression, but if it is not, it will be nice to have a more meaningful exception message. cc. @soumith

## To Reproduce

```sh
In [1]: import torch

In [2]: import numpy as np

In [3]: torch.tensor(np.array([1.]), dtype=torch.float32)  # Works fine
Out[3]: tensor([1.])

In [4]: torch.set_default_tensor_type('torch.cuda.FloatTensor')

In [5]: torch.tensor(np.array([1.]), dtype=torch.float32)   # Recent error
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-5b6e0594b346> in <module>()
----> 1 torch.tensor(np.array([1.]), dtype=torch.float32)

RuntimeError: device.has_value() ? device.value().type() == type.device_type() : other.device().type() == type.device_type() ASSERT FAILED at /opt/conda/conda-bld/pytorch-nightly_1543828543801/work/torch/csrc/utils/tensor_new.cpp:94, please report a bug to PyTorch.
```

## Environment

```sh
  $ python collect_env.py
Collecting environment information...
PyTorch version: 1.0.0.dev20181203
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 3.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.1.85
GPU models and configuration:
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080

Nvidia driver version: 390.30
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.3
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] numpy (1.15.1)
[pip] torch (1.0.0.dev20181203)
[pip] torchfile (0.1.0)
[pip] torchvision (0.2.1)
[conda] magma-cuda91              2.3.0                         1    pytorch
[conda] pytorch-nightly           1.0.0.dev20181203 py3.6_cuda9.0.176_cudnn7.4.1_0    pytorch
[conda] torchfile                 0.1.0                     <pip>
[conda] torchvision               0.2.1                     <pip>
```

### PR DESCRIPTION
**Pull request by t-vi**
This would solve the tracing problems of #13969.
Fixes: #14732

I would appreciate if this got good scrutiny before applied.

### CODE REVIEW

**Comment by t-vi** [comment]
I'll go through the CI failures to see if they're just renaming intermediates in the graphs or errors in the change itself.

**Comment by t-vi** [comment]
So the error for `TestJit.test_export_tensoroption_to` is because this
```python
import torch
def foo(x):
   return x.new_tensor(x[1]).cpu() + x
a = torch.randn([2])
y = torch.jit.trace(foo, (a,))
b = y(a)
print (torch.onnx._export_to_pretty_string(y, (a,), None, example_outputs=b))
```
now gives
```
ModelProto {
  producer_name: "pytorch"
  domain: ""
  doc_string: ""
  graph:
    GraphProto {
      name: "torch-jit-export"
      inputs: [{name: "x", type:Tensor dims: 2}]
      outputs: [{name: "2", type:Tensor dims: 2}]
      initializers: []
      nodes: [
        Node {type: "Constant", inputs: [], outputs: [1], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
        Node {type: "Add", inputs: [1,x], outputs: [2], attributes: []}
      ]
    }
  opset_import: [OperatorSetIdProto { domain: }],
}
```
rather than the following output for master:
```
ModelProto {
  producer_name: "pytorch"
  domain: ""
  doc_string: ""
  graph:
    GraphProto {
      name: "torch-jit-export"
      inputs: [{name: "x", type:Tensor dims: 2}]
      outputs: [{name: "7", type:Tensor dims: 2}]
      initializers: []
      nodes: [
        Node {type: "Constant", inputs: [], outputs: [1], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
        Node {type: "Gather", inputs: [x,1], outputs: [2], attributes: [{ name: 'axis', type: int, value: 0}]},
        Node {type: "Constant", inputs: [], outputs: [3], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
        Node {type: "Shape", inputs: [3], outputs: [4], attributes: []},
        Node {type: "Expand", inputs: [2,4], outputs: [5], attributes: []},
        Node {type: "Cast", inputs: [5], outputs: [6], attributes: [{ name: 'to', type: int, value: 11}]},
        Node {type: "Add", inputs: [6,x], outputs: [7], attributes: []}
      ]
    }
  opset_import: [OperatorSetIdProto { domain: }],
}
```
The JIT graph (`y.graph_for(a)`) looks perfectly good to me, but how can I be sure whether ONNX needs the reshape?

**Comment by t-vi** [comment]
The test failures are legit.

File: torch/csrc/utils/tensor_new.cpp [hunk 1]
```diff
@@ -206,45 +189,42 @@ Tensor internal_new_from_data(
...
   if (THPUtils_checkString(data)) {
     throw TypeError("new(): invalid data type '%s'", Py_TYPE(data)->tp_name);
   }
 
   if (THPVariable_Check(data)) {
     auto var = reinterpret_cast<THPVariable*>(data)->cdata;
-    auto type_inference_device_type = device_opt.has_value() ? device_opt->type()
-                                                             : torch::getDeviceType(var.type());
+    if (copy_variables) {
+      var = autograd::make_variable(var.data(), /*requires_grad=*/false);
```
Comment (gchanan) [inline torch/csrc/utils/tensor_new.cpp]: why do you need to do this?  Or rather, why didn't we need to do it before?

**Review by gchanan — COMMENTED** [review]
This looks fine to me, except that I don't know the answer to your onnx question.  It also makes me nervous to keep around both .to and dispatch_type_conversion -- can we combine them everywhere?

File: torch/csrc/utils/tensor_new.cpp (hunk 1 above)
Comment (t-vi) [reply torch/csrc/utils/tensor_new.cpp]: It's to disconnect it hard in the autograd graph when using `torch.tensor(...)`. My theory is that previously we used `Type.copy` which is autograd-breaking. Of course, one might ask whether it'd be cleaner to use `.detach()`.

**Comment by t-vi** [comment]
I'll look into replacing the other uses of `dispatch_type_conversion`.

File: test/expect/TestJit.test_export_tensoroption_to.expect [hunk 2]
```diff
@@ -6,16 +6,11 @@ ModelProto {
...
       initializers: []
       nodes: [
         Node {type: "Constant", inputs: [], outputs: [1], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
-        Node {type: "Gather", inputs: [x,1], outputs: [2], attributes: [{ name: 'axis', type: int, value: 0}]},
-        Node {type: "Constant", inputs: [], outputs: [3], attributes: [{ name: 'value', type: tensor, value:TensorProto shape: []}]},
-        Node {type: "Shape", inputs: [3], outputs: [4], attributes: []},
-        Node {type: "Expand", inputs: [2,4], outputs: [5], attributes: []},
-        Node {type: "Cast", inputs: [5], outputs: [6], attributes: [{ name: 'to', type: int, value: 11}]},
-        Node {type: "Add", inputs: [6,x], outputs: [7], attributes: []}
+        Node {type: "Add", inputs: [1,x], outputs: [2], attributes: []}
```
Comment (houseroad) [inline test/expect/TestJit.test_export_tensoroption_to.expect]: The changes on ONNX export look legit.

Comment (t-vi) [reply test/expect/TestJit.test_export_tensoroption_to.expect]: Thank you for weighing in!

**Comment by t-vi** [comment]
It seems the CI failures are all timeouts.
Other than that, I think things are working reasonably well.

**Comment by t-vi** [comment]
@ezyang  Should I rebase this for #14550 or should we just close it? I would greatly appreciate your advice.

**Comment by ezyang** [comment]
I'd definitely love to see this fixed, but let me first take a read and figure out if it would make sense to rebase or better to redo.

**Comment by ezyang** [comment]
I think this patch is worth rebasing; or, maybe not quite rebasing but using as a model to redo the patch. The hard part of actually landing this is appeasing JIT, I imagine.

**Comment by t-vi** [comment]
So I refreshed the patch, are those timeouts actually caused by the PR - they seem to happen too consistent...?

**Comment by t-vi** [comment]
When I run test_multiprocessing locally, I get an error "unsupported Storage type" before it hangs, so I guess I can debug this.

**Comment by soumith** [comment]
@ezyang can you help land this today, I'd like to see some linked issues fixed for release

**Comment by ezyang** [comment]
Looks like it just needs a slight expect file update. I'll run a local build of this.

**Comment by t-vi** [comment]
@ezyang Thank you for fixing this up!

**Pull request closed without merging**